darwin/Platform.c

```diff
@@ -357,9 +378,75 @@ void Platform_getFileDescriptors(double* used, double* max) {
 }
 
 bool Platform_getDiskIO(DiskIOData* data) {
-   // TODO
-   (void)data;
-   return false;
+   if (!IOKit_available)
+      return false;
+
+   io_iterator_t drive_list;
+
+   /* Get the list of all drives */
+   if (IOServiceGetMatchingServices(htop_main_port, IOServiceMatching("IOBlockStorageDriver"), &drive_list))
+      return false;
+
+   unsigned long long int read_sum = 0, write_sum = 0, timeSpend_sum = 0;
+   io_registry_entry_t drive;
+   while ((drive = IOIteratorNext(drive_list)) != 0) {
+      CFDictionaryRef properties = NULL;
+
+      /* Get the properties of this drive */
+      if (IORegistryEntryCreateCFProperties(drive, (CFMutableDictionaryRef*) &properties, kCFAllocatorDefault, 0)) {
+         IOObjectRelease(drive);
+         IOObjectRelease(drive_list);
+         return false;
+      }
+
+      if (properties != NULL) {
+         /* Get the statistics of this drive */
+         CFDictionaryRef statistics = (CFDictionaryRef) CFDictionaryGetValue(properties, CFSTR(kIOBlockStorageDriverStatisticsKey));
+
+         if (statistics != NULL) {
+            CFNumberRef number;
+            unsigned long long int value;
+
+            /* Get bytes read */
+            number = (CFNumberRef) CFDictionaryGetValue(statistics, CFSTR(kIOBlockStorageDriverStatisticsBytesReadKey));
+            if (number != 0) {
+               CFNumberGetValue(number, kCFNumberSInt64Type, &value);
+               read_sum += value;
+            }
+
+            /* Get bytes written */
+            number = (CFNumberRef) CFDictionaryGetValue(statistics, CFSTR(kIOBlockStorageDriverStatisticsBytesWrittenKey));
+            if (number != 0) {
+               CFNumberGetValue(number, kCFNumberSInt64Type, &value);
+               write_sum += value;
+            }
+
+            /* Get total read time (in ns) */
+            number = (CFNumberRef) CFDictionaryGetValue(statistics, CFSTR(kIOBlockStorageDriverStatisticsTotalReadTimeKey));
+            if (number != 0) {
+               CFNumberGetValue(number, kCFNumberSInt64Type, &value);
+               timeSpend_sum += value;
+            }
+
+            /* Get total write time (in ns) */
+            number = (CFNumberRef) CFDictionaryGetValue(statistics, CFSTR(kIOBlockStorageDriverStatisticsTotalWriteTimeKey));
+            if (number != 0) {
+               CFNumberGetValue(number, kCFNumberSInt64Type, &value);
+               timeSpend_sum += value;
+            }
+         }
+         CFRelease(properties);
+      }
+      IOObjectRelease(drive);
+   }
+   data->totalBytesRead = read_sum;
+   data->totalBytesWritten = write_sum;
```

Comment:
No need to cramp everything together …
```suggestion
         }

         CFRelease(properties);
      }

      IOObjectRelease(drive);
   }

   data->totalBytesRead = read_sum;
   data->totalBytesWritten = write_sum;
```